consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftController.java

```diff
@@ -68,20 +77,24 @@ public IbftController(
       final IbftFinalState ibftFinalState,
       final IbftBlockHeightManagerFactory ibftBlockHeightManagerFactory,
       final Gossiper gossiper,
-      final Map<Long, List<Message>> futureMessages) {
+      final Map<Long, List<Message>> futureMessages,
+      final MessageTracker messageTracker) {
     this.blockchain = blockchain;
     this.ibftFinalState = ibftFinalState;
     this.ibftBlockHeightManagerFactory = ibftBlockHeightManagerFactory;
     this.futureMessages = futureMessages;
     this.gossiper = gossiper;
+    this.messageTracker = messageTracker;
   }
 
   public void start() {
     startNewHeightManager(blockchain.getChainHeadHeader());
   }
 
   public void handleMessageEvent(final IbftReceivedMessageEvent msg) {
-    handleMessage(msg.getMessage());
+    if (!messageTracker.hasSeenMessage(msg.getMessage().getData())) {
```

Comment:
Maybe this should be reconsidered - i.e. it should say "If we have SEEN this message before, do not handle it" - whereas it currently says "If I have SENT/GOSSIPED this message before, do not handle it" ... I believe our requirement/concept should be the first of these.

Reply:
Thinking about it more your right. I'm preventing sent messages from being processed but it's previously received messages that we need to filter out.

Reply:
Added seperate buffer for seen messages in the IbftController